### PR TITLE
Create redundant_facet_removal_backup.py

### PR DESCRIPTION
**Pull request by nitishmalang**
pull request

### CODE REVIEW

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_backup.py]: plz change this to:
model_iter = solve_lp_with_different_objectives(
                                model_iter.copy(), objective_function
                            )

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_backup.py]: similar to the previous call

File: dingo/backup/redundant_facet_removal_backup.py
Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_backup.py]: ok

**Comment by TolisChal** [comment]
Thanks @nitishmalang! I'm ok with merging. Let's wait for a second review (@vissarion ?)

**Review by TolisChal — APPROVED** [review]

**Comment by nitishmalang** [comment]
ok thanks

**Review by TolisChal — APPROVED** [review]

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_backup.py]: Could you plz remove fast_fba() and fast_fva() from this file since they are irrelevant?

**Review by nitishmalang — COMMENTED** [review]
done

**Review by TolisChal — APPROVED** [review]

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_backup.py]: This has to be:
model_iter = solve_lp_with_different_objectives(
                            model_iter.copy(), objective_function_max
                        )

Plz change this wherever it has

Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_backup.py]: ok

**Review by nitishmalang — COMMENTED** [review]
did it

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_backup.py]: Could you plz also remove the function fast_inner_ball() from this file?

Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_backup.py]: okay

Comment (TolisChal) [reply dingo/backup/redundant_facet_removal_backup.py]: Thanks!

**Review by nitishmalang — COMMENTED** [review]
.

**Review by TolisChal — APPROVED** [review]

File: dingo/backup/redundant_facet_removal_backup.py
Comment (vissarion) [inline dingo/backup/redundant_facet_removal_backup.py]: This is to replace `fast_remove_redundant_facets` in https://github.com/GeomScale/dingo/blob/develop/dingo/gurobi_based_implementations.py#L312C4-L312C4

File: dingo/backup/redundant_facet_removal_backup.py
Comment (vissarion) [inline dingo/backup/redundant_facet_removal_backup.py]: This is to replace `update_model` in https://github.com/GeomScale/dingo/blob/develop/dingo/gurobi_based_implementations.py#L277

File: dingo/backup/redundant_facet_removal_backup.py
Comment (vissarion) [inline dingo/backup/redundant_facet_removal_backup.py]: this is the main change here (and one call below) that replaces the `update_model` and `optimize()` calls.

File: dingo/backup/update_gurobi_model_backup.py
Comment (vissarion) [inline dingo/backup/update_gurobi_model_backup.py]: I do not understand the purpose of this function. Is this to be called by `fast_remove_redundant_facets` or replace `update_model` in certain cases?

**Review by vissarion — APPROVED** [review]
Thanks @nitishmalang! I am in general OK with merging this code.

I have a few comments that could be integrated in the code. Please insert them as comments if you agree. 

Also I propose to use a different name for  the directory than `backup`. Maybe `experimental_gurobi_functions` and remove the term `backup` from all file names. This is not a backup it is the opposite it is something new, but experimental, that is why we are not integrating directly into the codebase, but it has to be tested.

File: dingo/backup/redundant_facet_removal_backup.py
Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_backup.py]: yes this one is to be replaced

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [reply dingo/backup/redundant_facet_removal_backup.py]: No, this is to be deleted. @nitishmalang could you plz delete this function?

File: dingo/backup/redundant_facet_removal_backup.py
Comment (TolisChal) [reply dingo/backup/redundant_facet_removal_backup.py]: This function can be used to change only the objective function of a model. So, this is to be used in the loop 
`for i in indices:`

Inside this loop we have a certain set of constraints and we check for redundant facets.

File: dingo/backup/update_gurobi_model_backup.py
Comment (TolisChal) [reply dingo/backup/update_gurobi_model_backup.py]: This function is to be used in the loop:
`while removed > 0 or offset > 0:`
and before starting the loop:
`for i in indices:`

When we remove a facet the set of constraints changes, so we have to update the model (change the matrices A,Aeq,b,beq and the bounds). Then, we have to run `for i in indices:` and go over all the active coordinates to check for new redundant facets.

File: dingo/backup/redundant_facet_removal_backup.py
Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_backup.py]: ok @TolisChal

File: dingo/backup/redundant_facet_removal_experimental.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_experimental.py]: Could you plz delete this function since it is not needed here?

File: dingo/backup/redundant_facet_removal_experimental.py
Comment (TolisChal) [inline dingo/backup/redundant_facet_removal_experimental.py]: Can you add a line before ` for i in indices:` to update the model?
i.e.,
update_model_constraints_and_bounds(model_iter, Aeq_sparse, beq, A_sparse, [val], lb, ub)

Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_experimental.py]: ok

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (TolisChal) [inline dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: Could you plz delete all these empty lines?

Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: ok

File: dingo/experimental_gurobi_functions/update_gurobi_model_experiment.py
Comment (TolisChal) [inline dingo/experimental_gurobi_functions/update_gurobi_model_experiment.py]: Could you plz move this function into the file `dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py` and delete this file?

Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/update_gurobi_model_experiment.py]: ok @TolisChal

File: dingo/backup/redundant_facet_removal_experimental.py
Comment (TolisChal) [reply dingo/backup/redundant_facet_removal_experimental.py]: Is this resolved indeed? I think this is not implemented yet.

**Review by nitishmalang — COMMENTED** [review]
updated @TolisChal

File: dingo/backup/redundant_facet_removal_experimental.py
Comment (nitishmalang) [reply dingo/backup/redundant_facet_removal_experimental.py]: done this

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (TolisChal) [inline dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: plz fix the spaces in this line. That will give a runtime error

Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: ok fixing this

**Review by nitishmalang — COMMENTED** [review]
.

**Review by nitishmalang — COMMENTED** [review]
updated

**Review by nitishmalang — COMMENTED** [review]
indentation change is done

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (TolisChal) [inline dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: This is written twice. If there is no reason for this, plz delete one of those two

Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: ok doing it

**Review by TolisChal — APPROVED** [review]

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (vissarion) [inline dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: This call seems wrong. Shouldn't it be  `update_model_constraints_and_bounds(model_iter, Aeq_sparse, beq, A_sparse, [val], lb, ub)`?

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (vissarion) [inline dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py:65]: is this needed? Since you will update the objective function in the next call.

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (vissarion) [inline dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py:73]: optimization todo: you could pass a numpy array here (instead of a list of coefficients) and update the objective in one call without a for loop.

It is ok if you do not implement this in this PR but please add it as a comment

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py]: ok did it

**Review by nitishmalang — COMMENTED** [review]
.

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py:65]: the clearing of the existing objective function is indeed a precautionary step to ensure no remnants of the old objective function affect the subsequent update. It's a safety measure for robustness.

**Review by nitishmalang — COMMENTED** [review]
updated

File: dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py
Comment (nitishmalang) [reply dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py:73]: updated it

**Review by TolisChal — APPROVED** [review]

**Pull request merged**
